### PR TITLE
feat: support flat tree dependencies structure

### LINKED ISSUE #403: Flat tree if npm 3 engine
This is a feature request.
Some projects only work with npm 3 due to the need of a flat tree.
My suggestion is to check the npm engine in package.json and if >=3, generate a flat tree and/or add a flag --flat-tree to the command line.

EDIT:

```
git clone https://github.com/zkochan/pnpm-issue394.git
cd pnpm-issue394
pnpm i
node_modules/.bin/ng build
```
#394 has to be solved first. Then you will have this error:

```
 94% asset optimizationError in bail mode: [default] C:\Users\aecz\pnpm-issue394\src\app\app.component.ts:5:12
Cannot find name 'require'.
```

Package @typings/node is expected at the top level of node_modules but is in fact in protractor submodule.

### PR DESCRIPTION
**Pull request by zkochan**
Close #403

BREAKING CHANGE:

Store divided into `flat/` and `nested/`

### CODE REVIEW

**Comment by zkochan** [comment]
This is a breaking change, so I also included another breaking change. I changed store.json to store.yaml. yarn uses the yaml format as well, in yarn.lock and I believe we will support the lock file as well in the future, so lets have consistency. yaml is more readable anyway

**Comment by iamstarkov** [comment]
> Store divided into flat/ and nested/

what is default store? i mean how do i access binaries or node_modules installed by pnpm?

**Comment by zkochan** [comment]
Currently it is either in `~/pnpm/.store/` if global store is used

or in `node_modules/.store` if a local store is used.

After the change there will be two subfolders. One for packages installed using flat tree and another one for packages installed using nested dependencies structure

**Comment by iamstarkov** [comment]
wont it break everything? 'cause no one packages expect node_modules to be separated in two folders

**Comment by zkochan** [comment]
every package will use either the flat store or the nested store

these two types won't be mixed

File: src/api/getContext.ts [hunk 1]
```diff
@@ -7,41 +7,40 @@ import writeJson from '../fs/writeJson'
...
 }
 
 export default async function (opts: StrictPnpmOptions): Promise<PnpmContext> {
   const pkg = await (opts.global ? readGlobalPkg(opts.globalPath) : readPkgUp({ cwd: opts.cwd }))
   const root = normalizePath(pkg.path ? path.dirname(pkg.path) : opts.cwd)
-  const store = resolveStorePath(opts.storePath, root)
-  const storeJson = readStoreJson(store)
-  if (storeJson) {
-    failIfNotCompatible(storeJson.pnpm)
+  const storePath = path.join(resolveStorePath(opts.storePath, root), opts.flatTree ? 'flat' : 'nested')
```
Comment (zkochan) [inline src/api/getContext.ts]: @iamstarkov look. One of the stores will be used. They won't get mixed

**Comment by zkochan** [comment]
When the store is inside node_modules, creating an additional subdirectory is redundant because node_modules will be either flat or nested. Hence, I will remove the additional nesting in dedicated stores and will add an additional property to `store.yaml` to identify the store type: nested or flat.

File: src/api/extendOptions.ts [hunk 2]
```diff
@@ -29,5 +30,9 @@ function getDefaultStorePath () {
 }
 
 export default (opts?: PnpmOptions): StrictPnpmOptions => {
+  opts = opts || {}
+  if (opts.flatTree === true && !preserveSymlinks) {
+    throw new TypeError('Flat tree is supported only on Node.js >= 6.3.0')
```
Comment (iamstarkov) [inline src/api/extendOptions.ts]: if expression is little relevant to thrown error, thus it makes it and solution to it a bit obscure. maybe:

```
`--preserve-symlinks` and so 'Flat tree' are not supported on your system, make sure you are running on Node ≽ 6.3.0
```

Comment (iamstarkov) [reply src/api/extendOptions.ts]: and i think its not `TypeError`

Comment (iamstarkov) [reply src/api/extendOptions.ts]: Though, do we want to scream and stop work or warn and continue with non-flat tree?

File: src/fs/yamlfs.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import fs = require('fs')
+import yaml = require('js-yaml')
+
+export function read <T>(yamlPath: string): T {
+  return yaml.safeLoad(fs.readFileSync(yamlPath, 'utf8'))
+}
+
+export function write <T>(yamlPath: string, yamlObj: T) {
+  fs.writeFileSync(yamlPath, yaml.safeDump(yamlObj), 'utf8')
```
Comment (iamstarkov) [inline src/fs/yamlfs.ts]: are you sure about `sync`?

File: test/install.ts [hunk 4]
```diff
@@ -290,7 +290,7 @@ test('shrinkwrap compatibility', async function (t) {
   await installPkgs(['rimraf@2.5.1'], testDefaults())
 
   return new Promise((resolve, reject) => {
-    const proc = crossSpawn.spawn('npm', ['shrinkwrap'])
+    const proc = crossSpawn.spawn('npm', ['shrinkwrap'], {stdio: 'inherit'})
```
Comment (iamstarkov) [inline test/install.ts]: cool stuff. though what is `stdio: inherit` for?

File: test/install.ts [hunk 5]
```diff
@@ -119,7 +119,7 @@ test('overwriting (magic-hook@2.0.0 and @0.1.0)', async function (t) {
   prepare()
   await installPkgs(['magic-hook@2.0.0'], testDefaults())
 
-  const flattenPathInStore = path.join(process.cwd(), 'node_modules', '.store', 'flatten@1.0.2')
+  const flattenPathInStore = path.join(process.cwd(), 'node_modules/.store/flatten@1.0.2')
```
Comment (iamstarkov) [inline test/install.ts:122]: i believe all these tests will stop to work on windows

**Comment by iamstarkov** [comment]
if in the future we want to use `.store.yaml` the same way as `.yarn.lock` then we might want to move it out from node_modules now

**Comment by iamstarkov** [comment]
> When the store is inside node_modules, creating an additional subdirectory is redundant because node_modules will be either flat or nested. Hence, I will remove the additional nesting in dedicated stores and will add an additional property to store.yaml to identify the store type: nested or flat.

Before this i was confused and couldnt get how it would work, now im fine about this part

File: test/install.ts (hunk 4 above)
Comment (zkochan) [reply test/install.ts]: oops, no need in that... that was for debug purposes because the test was failing

File: test/install.ts (hunk 5 above)
Comment (zkochan) [reply test/install.ts:122]: tests on appveyor seem to work... probably path.join is smart enough

File: src/api/extendOptions.ts (hunk 2 above)
Comment (zkochan) [reply src/api/extendOptions.ts]: I think an error is better. It is a new property not available in npm anyway

File: src/fs/yamlfs.ts (hunk 3 above)
Comment (zkochan) [reply src/fs/yamlfs.ts]: I'll change it to async

**Comment by zkochan** [comment]
> if in the future we want to use .store.yaml the same way as .yarn.lock then we might want to move it out from node_modules now

store.yaml is for the store. yarn.lock is for the package. Also store.yaml does not influence what is installed, so it has different purposes.

A lockfile we'll have to implement from scratch, probably.

**Comment by iamstarkov** [comment]
> store.yaml is for the store. yarn.lock is for the package

store is list of deps with exact versions, lock is kinda the same. You sort of oppose one to another, but i think they are interrelated with each other.

> Also store.yaml does not influence what is installed, so it has different purposes

right now its correct, but move it out and consumer would need to run `pnpm i` first time to get all modules` into shared store via network and to get them linked into node_modules, second time its just to link them again (wo/ network overhead)

**Comment by zkochan** [comment]
hm, makes sense, it can be used as a lock!

but it contains all the deps of all the packages that use the store. It would work with a dedicated store but a shared store's store.yaml would have a lot of redundant staff..

**Comment by iamstarkov** [comment]
> but it contains all the deps of all the packages that use the store. It would work with a dedicated store but a shared store's store.yaml would have a lot of redundant staff..

I believe it includes all deps as a graph of dependents and dependencies, without any redundant things, because all of that you need to perform `install/remove --save-*` safely

**Comment by zkochan** [comment]
you are right. But 1 store can be used by several packages and all these graphs will be in one store.yaml

so currently we can't just move this file to one of the packages

File: src/fs/modulesController.ts [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+export type Modules = {
+  storePath: string,
+}
+
+export async function read (modulesPath: string): Promise<Modules | null> {
+  const modulesYamlPath = path.join(modulesPath, modulesFileName)
+  try {
+    return await readYaml<Modules>(modulesYamlPath)
+  } catch (err) {
+    return null
```
Comment (iamstarkov) [inline src/fs/modulesController.ts:23]: wont returning `null` instead of actual `err` make it hard to debug?

**Comment by iamstarkov** [comment]
> you are right. But 1 store can be used by several packages and all these graphs will be in one store.yaml
> 
> so currently we can't just move this file to one of the packages

you are right about it, i was thinking about "personal" store for each project `${project}/node_modules/.store.yaml`. So i guess we should have two stores: one global and one for each project

**Comment by iamstarkov** [comment]
and i think local stores can not have whole graph, but just upper graph points as kind of entry points to global store's whole graph

**Comment by zkochan** [comment]
> So i guess we should have two stores: one global and one for each project

I think we have to use some other naming for the one which is in the project

**Comment by zkochan** [comment]
>  and i think local stores can to not have whole graph, but just upper graph points as kind of entry points to global store's whole graph

so this graph will be both in the global store and locally in the project?

**Comment by zkochan** [comment]
I think we have to create a separate issue for this because it there's a lot to discuss

**Comment by iamstarkov** [comment]
minor suggestion `.pnpm.lock`, so it and `.yarn.lock` can be ignored by `.*.lock` in ignore files

**Comment by iamstarkov** [comment]
> so this graph will be both in the global store and locally in the project?

nej, i see global store's graph as graph from exact versions to another exact versions. and local store's as a list of exact versions of project's direct deps, which can be used to define whole graph with help of global store. got it?

**Comment by zkochan** [comment]
I think I did got you. You suggest to save entry points in the lock file. But what is the value of such a file if it is not usable without the store.yaml. On the other hand, the same can be achieved without using the lock file. We already have the single entry point, which is the path to the package.

**Comment by iamstarkov** [comment]
lets discuss it later

**Pull request merged**